flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/results/serde/LogicalTypeJsonDeserializer.java

```diff
@@ -0,0 +1,219 @@
+        switch (typeRoot) {
+            case BOOLEAN:
+                return new BooleanType(isNullable);
+            case TINYINT:
+                return new TinyIntType(isNullable);
+            case SMALLINT:
+                return new SmallIntType(isNullable);
+            case INTEGER:
+                return new IntType(isNullable);
+            case BIGINT:
+                return new BigIntType(isNullable);
+            case FLOAT:
+                return new FloatType(isNullable);
+            case DOUBLE:
+                return new DoubleType(isNullable);
+            case DATE:
+                return new DateType(isNullable);
+            case NULL:
+                return new NullType();
+            case CHAR:
+            case VARCHAR:
+            case BINARY:
+            case VARBINARY:
+                return deserializeLengthFieldType(typeRoot, logicalTypeNode).copy(isNullable);
+            case DECIMAL:
+                return new DecimalType(
+                        isNullable,
+                        logicalTypeNode.get(FIELD_NAME_PRECISION).asInt(),
+                        logicalTypeNode.get(FIELD_NAME_SCALE).asInt());
+            case TIMESTAMP_WITHOUT_TIME_ZONE:
+            case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+                return deserializeTimestamp(typeRoot, logicalTypeNode).copy(isNullable);
+            case MAP:
+                return deserializeMap(logicalTypeNode).copy(isNullable);
+            case ARRAY:
+            case MULTISET:
+                return deserializeCollection(typeRoot, logicalTypeNode).copy(isNullable);
+            case ROW:
+                return deserializeRow(logicalTypeNode).copy(isNullable);
+            case RAW:
+                return deserializeRaw(logicalTypeNode).copy(isNullable);
+            default:
+                throw new TableException("Unsupported type root: " + typeRoot);
+        }
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // Helper methods for some complex types
+    // --------------------------------------------------------------------------------------------
+
+    private LogicalType deserializeLengthFieldType(
+            LogicalTypeRoot typeRoot, JsonNode logicalTypeNode) {
+        int length = logicalTypeNode.get(FIELD_NAME_LENGTH).asInt();
+        switch (typeRoot) {
+            case CHAR:
+                return length == 0 ? CharType.ofEmptyLiteral() : new CharType(length);
+            case VARCHAR:
+                return length == 0 ? VarCharType.ofEmptyLiteral() : new VarCharType(length);
+            case BINARY:
+                return length == 0 ? BinaryType.ofEmptyLiteral() : new BinaryType(length);
+            case VARBINARY:
+                return length == 0 ? VarBinaryType.ofEmptyLiteral() : new VarBinaryType(length);
+            default:
+                throw new TableException("Logical type with attribute 'length' expected.");
```

Comment:
"Can't convert json '%s' to the logical type %s."

BTW, I think it's better to throw SqlGatewayException here.